app/src/main/java/com/nutomic/syncthingandroid/service/ReceiverManager.java

```diff
@@ -0,0 +1,68 @@
+package com.nutomic.syncthingandroid.service;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class ReceiverManager {
+
+    private static final String TAG = "ReceiverManager";
+
+    private static List<BroadcastReceiver> mReceivers = new ArrayList<BroadcastReceiver>();
+
+    public static void registerReceiver(Context context, BroadcastReceiver receiver, IntentFilter intentFilter) {
+        mReceivers.add(receiver);
+        context.registerReceiver(receiver, intentFilter);
+        Log.v(TAG, "Registered receiver: " + receiver + " with filter: " + intentFilter);
+    }
+
+    public static boolean isReceiverRegistered(BroadcastReceiver receiver) {
+        return mReceivers.contains(receiver);
+    }
+
+    public static void unregisterReceiver(Context context, BroadcastReceiver receiver) {
+        if (context == null) {
+            Log.e(TAG, "unregisterReceiver: context is null");
+            return;
+        }
+        if (isReceiverRegistered(receiver)) {
+            try {
+                context.unregisterReceiver(receiver);
+                Log.v(TAG, "Unregistered receiver: " + receiver);
+            } catch(IllegalArgumentException e) {
+                // We have to catch the race condition a registration is still pending in android
+                // according to https://stackoverflow.com/a/3568906
+                Log.w(TAG, "unregisterReceiver(" + receiver + ") threw IllegalArgumentException");
+            }
+            mReceivers.remove(receiver);
+        }
+    }
+
+    public static void unregisterAllReceivers(Context context) {
+        if (context == null) {
+            Log.e(TAG, "unregisterReceiver: context is null");
+            return;
+        }
+        Iterator<BroadcastReceiver> iter = mReceivers.iterator();
+        while (iter.hasNext()) {

```

Comment:
Note - this has to be an iterator and not a foreach(). If using foreach, elements mustn't be removed which we need to do. See foreach() crash that got solved by using the iterator.
![image](https://user-images.githubusercontent.com/16361913/40931469-81367cf8-682b-11e8-9bf0-64aab10506e3.png)